Implement Kronecker product of Matrices

raineszm opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

This pull request adds a `MatrixExpr` class which represents the [Kronecker product](https://en.wikipedia.org/wiki/Kronecker_product) of matrices.

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
This PR can be seen as a stepping stone to resolving #10771, #10734 , and  #11832 but these commits do not touch the `physics.quantum.tensorproduct` module.


#### Brief description of what is fixed or changed

Add in this PR is the `sympy.matrices.expressions.kronecker` module which exports

* The `kronecker_product()` function, *which attempts to take the Kronecker product of its arguments*
* The `KroneckerProduct` class, *which represents and unevaluated Kronecker product of matrices*
* The `combine_kronecker()` function, *which attempts to combine operations on Kronecker products into Kronecker products of those operations*

The only changes to existing modules in this PR are to add KroneckerProduct to the printing systems and to import the aforementioned functions and class into the `sympy.matrices.expressions` module.

raineszm committed:
Create KroneckerProduct MatrixExpr class

Implemented a class that represents a the KroneckerProduct of Matrices and a
method kronecker_product() with constructs such an object and then evaluates it
if possible.

raineszm committed:
Move KroneckerProduct printing to printer classes

KroneckerProduct previously used the old style printing architecture i.e.
_sympystr, _latex methods, etc.  This functionality is now implemented directly
in the printer classes as is the case for other builtin Matrix expressions.

raineszm committed:
Add combine_kronecker method

Expressions containing operations on KroneckerProducts can now be
simplified by grouping the KronekcerProducts together

raineszm committed:
matrices: Add expand kroneckerproduct hint

This will expand a KroneckerProduct of sums into a sum of
KroneckerProducts

raineszm committed:
Convert KroneckerProduxct of Identities to Identity

raineszm committed:
Test KroneckerProduct negatives and inverses

asmeurer commented on sympy/printing/str.py:
```diff
@@ -311,6 +311,10 @@ def _print_HadamardProduct(self, expr):
             for arg in expr.args])
 
+    def _print_KroneckerProduct(self, expr):
+        return 'x'.join([self.parenthesize(arg, precedence(expr))
```
The str printer should be a recreatable form. Pretty printing should go in the pretty printer.

raineszm replied:
Ah, my apologies. I was using the printing of other classes in the `sympy.matrix.expressions` module as a guideline. I'll remove the method in the str printer.

asmeurer commented on sympy/printing/pretty/pretty.py:
```diff
@@ -857,6 +857,15 @@ def _print_HadamardProduct(self, expr):
+        if self._use_unicode:
+            delim = u' \N{N-ARY CIRCLED TIMES OPERATOR} '
+        else:
+            delim = 'x'
```
Shouldn't there be a space for the ascii printer too?

raineszm replied:
Yes, there probably should be.

asmeurer commented:
CC @Upabjojr

raineszm committed:
Remove KroneckerProduct from str printer

raineszm committed:
Modify pretty printing of KroneckerProduct

Add whitespace to the ascii representation of the KroneckerProduct to improve
readiblity and match better with the unicode case.

jksuom commented:
Kronecker product of matrices, which is nowadays usually called tensor product, is also constructed in `sympy.physics.quantum.tensorproduct`.  Could that be made use of here?

Upabjojr commented:
I believe we should not use the term tensor product for Kronecker product. I would expect the tensor product of two matrices to create a four dimensional array, while the Kronecker product acts like in this PR.

This PR appears to be a step in the right direction.

Upabjojr commented:
I like this PR. I'm OK to merge this. What do you think?
